apps/enc.c

```diff
@@ -418,15 +426,26 @@ int enc_main(int argc, char **argv)
                     BIO_printf(bio_err, "bad magic number\n");
                     goto end;
                 }
-
                 sptr = salt;
             }
 
-            if (!EVP_BytesToKey(cipher, dgst, sptr,
-                                (unsigned char *)str,
-                                str_len, 1, key, iv)) {
-                BIO_printf(bio_err, "EVP_BytesToKey failed\n");
-                goto end;
+            if (pbkdf2 == 1) {
+                if (!PKCS5_PBKDF2_HMAC(str, str_len, sptr, sizeof(salt), iter,
+                                       dgst,
+                                       EVP_CIPHER_key_length(cipher), key)) {
+                    BIO_printf(bio_err, "PKCS5_PBKDF2_HMAC failed\n");
+                    goto end;
+                }
+                /* should we reject if no IV is given ? */
```

Comment:
Hmmm, that's not my understanding. PBKDF2 effectively operates as a PRF, and there are no bounds on the amount of material that you can read from it.  RFC2898 is a little unclear on this point, but it does say this:
````
4. Encrypt the message M with the underlying encryption scheme
         under the derived key DK to produce a ciphertext C. (This step
         may involve selection of parameters such as an initialization
         vector and padding, depending on the underlying scheme.)
````
I read that as saying that the parameters may be come from DK too.  The PBES1 example is more explicit
````
 3. Separate the derived key DK into an encryption key K consisting
         of the first eight octets of DK and an initialization vector IV
         consisting of the next eight octets:
````

So, my reading of RFC2898 would allow this.